### PR TITLE
Game command args

### PR DESCRIPTION
**Pull request by reinaldorauch**
Implementing the game command arg structs

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Could you rebase now that #1905 is merged.

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 1]
```diff
@@ -719,6 +778,47 @@ namespace OpenLoco::GameCommands
...
 
+    struct ChangeCompanyColourSchemeArgs
+    {
+        ChangeCompanyColourSchemeArgs() = default;
+        explicit ChangeCompanyColourSchemeArgs(const registers& regs)
+            : companyId(CompanyId(regs.dl))
+            , isPrimary(regs.ah)
+            , value(regs.al)
+            , colourType(regs.cl)
+            , setColourMode() // TODO: Convert into another command for enabling sub coloring
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:789]: just do setColourMode(regs.dh)

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 2]
```diff
@@ -719,6 +778,47 @@ namespace OpenLoco::GameCommands
         doCommand(GameCommand::changeCompanyColourScheme, regs);
     }
 
+    struct ChangeCompanyColourSchemeArgs
+    {
+        ChangeCompanyColourSchemeArgs() = default;
+        explicit ChangeCompanyColourSchemeArgs(const registers& regs)
+            : companyId(CompanyId(regs.dl))
+            , isPrimary(regs.ah)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:786]: this should move to the body of the constructor and do `if (setColourMode == 0) { isPrimary = regs.ah == 0; }`

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 3]
```diff
@@ -719,6 +778,47 @@ namespace OpenLoco::GameCommands
...
+            registers regs;
+
+            regs.cl = colourType;           // vehicle type or main
+            regs.dh = setColourMode;        // [ 0, 1 ] -- 0 = set colour, 1 = toggle enabled/disabled;
+            regs.dl = enumValue(companyId); // company id
+
+            if (setColourMode == 0)
+            {
+                // cl is divided by 2 when used
+                regs.ah = isPrimary; // [ 0, 1 ] -- primary or secondary palette
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:810]: once you switch to a bool for isPrimary do `regs.ah = isPrimary ? 0 : 1`

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 4]
```diff
@@ -719,6 +778,47 @@ namespace OpenLoco::GameCommands
         doCommand(GameCommand::changeCompanyColourScheme, regs);
     }
 
+    struct ChangeCompanyColourSchemeArgs
+    {
+        ChangeCompanyColourSchemeArgs() = default;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:783]: dont forget to add  `static constexpr auto command = GameCommand::changeCompanyColourScheme` goes for a good few of these structs.

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 5]
```diff
@@ -842,6 +978,39 @@ namespace OpenLoco::GameCommands
...
+    {
+        static constexpr auto command = GameCommand::raiseLand;
+
+        RaiseLandArgs() = default;
+        explicit RaiseLandArgs(const registers& regs)
+            : centre(World::Pos2(regs.ax, regs.cx))
+            , pointA(World::Pos2(regs.edx, regs.ebp))             // lower bits
+            , pointB(World::Pos2(regs.edx >> 16, regs.ebp >> 16)) // higher bits
+            , corner(regs.di)
+            , flags(regs.bl)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:991]: don't add flags here.

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 6]
```diff
@@ -842,6 +978,39 @@ namespace OpenLoco::GameCommands
...
+        {
+        }
+
+        World::Pos2 centre;
+        World::Pos2 pointA;
+        World::Pos2 pointB;
+        uint16_t corner;
+        uint8_t flags;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:999]: remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 7]
```diff
@@ -842,6 +978,39 @@ namespace OpenLoco::GameCommands
...
+
+        explicit operator registers() const
+        {
+            registers regs;
+            regs.ax = centre.x;
+            regs.cx = centre.y;
+            regs.edx = pointB.x << 16 | pointA.x;
+            regs.ebp = pointB.y << 16 | pointA.y;
+            regs.bl = flags;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1008]: remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 8]
```diff
@@ -855,6 +1024,39 @@ namespace OpenLoco::GameCommands
...
+    {
+        static constexpr auto command = GameCommand::lowerLand;
+
+        LowerLandArgs() = default;
+        explicit LowerLandArgs(const registers& regs)
+            : centre(World::Pos2(regs.ax, regs.cx))
+            , pointA(World::Pos2(regs.edx, regs.ebp))             // lower bits
+            , pointB(World::Pos2(regs.edx >> 16, regs.ebp >> 16)) // higher bits
+            , corner(regs.di)
+            , flags(regs.bl)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1037]: as above remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 9]
```diff
@@ -842,6 +978,39 @@ namespace OpenLoco::GameCommands
...
     }
 
+    struct RaiseLandArgs
+    {
+        static constexpr auto command = GameCommand::raiseLand;
+
+        RaiseLandArgs() = default;
+        explicit RaiseLandArgs(const registers& regs)
+            : centre(World::Pos2(regs.ax, regs.cx))
+            , pointA(World::Pos2(regs.edx, regs.ebp))             // lower bits
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:988]: you'll need to mask off the higher bytes or use regs.dx and regs.bp

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 10]
```diff
@@ -855,6 +1024,39 @@ namespace OpenLoco::GameCommands
...
     }
 
+    struct LowerLandArgs
+    {
+        static constexpr auto command = GameCommand::lowerLand;
+
+        LowerLandArgs() = default;
+        explicit LowerLandArgs(const registers& regs)
+            : centre(World::Pos2(regs.ax, regs.cx))
+            , pointA(World::Pos2(regs.edx, regs.ebp))             // lower bits
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1034]: as above needs masking or using regs.dx, regs.bp

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 11]
```diff
@@ -868,6 +1070,39 @@ namespace OpenLoco::GameCommands
...
     }
 
+    struct LowerRaiseLandMountainArgs
+    {
+        static constexpr auto command = GameCommand::lowerRaiseLandMountain;
+
+        LowerRaiseLandMountainArgs() = default;
+        explicit LowerRaiseLandMountainArgs(const registers& regs)
+            : centre(World::Pos2(regs.ax, regs.cx))
+            , pointA(World::Pos2(regs.edx, regs.ebp))             // lower bits
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1080]: and here

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 12]
```diff
@@ -868,6 +1070,39 @@ namespace OpenLoco::GameCommands
...
+            , pointB(World::Pos2(regs.edx >> 16, regs.ebp >> 16)) // higher bits
+            , corner(regs.di)
+            , flags(regs.bl)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1083]: and here

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 13]
```diff
@@ -880,6 +1115,33 @@ namespace OpenLoco::GameCommands
...
     }
 
+    struct RaiseWaterArgs
+    {
+        static constexpr auto command = GameCommand::raiseWater;
+        RaiseWaterArgs() = default;
+        explicit RaiseWaterArgs(const registers& regs)
+            : pointA(regs.ax, regs.cx)
+            , pointB(regs.di, regs.bp)
+            , flags(regs.bl)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1125]: remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 14]
```diff
@@ -905,6 +1194,36 @@ namespace OpenLoco::GameCommands
...
+            , bufferIndex(regs.ax)
+        {
+            memcpy(newName, &regs.edx, 4);
+            memcpy(newName + 4, &regs.ebp, 4);
+            memcpy(newName + 8, &regs.edi, 4);
+        }
+
+        CompanyId companyId;
+        uint16_t bufferIndex;
+        char newName[36 * 16];
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1213]: 37 chars for a this not 36*16

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 15]
```diff
@@ -918,6 +1237,37 @@ namespace OpenLoco::GameCommands
...
+    struct ChangeCompanyOwnerNameArgs
+    {
+        static constexpr auto command = GameCommand::changeCompanyOwnerName;
+
+        ChangeCompanyOwnerNameArgs() = default;
+        explicit ChangeCompanyOwnerNameArgs(const registers& regs)
+            : companyId(CompanyId(regs.cx))
+            , bufferIndex(regs.ax)
+        {
+            memcpy(newName, &regs.ecx, 4);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1249]: this looks wrong.

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 16]
```diff
@@ -918,6 +1237,37 @@ namespace OpenLoco::GameCommands
...
+            memcpy(newName + 4, &regs.edx, 4);
+            memcpy(newName + 8, &regs.ebp, 4);
+            memcpy(newName + 12, &regs.edi, 4);
+        }
+
+        CompanyId companyId;
+        uint16_t bufferIndex;
+        char newName[36 * 16];
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1257]: 37

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 17]
```diff
@@ -990,6 +1340,36 @@ namespace OpenLoco::GameCommands
...
 
+    struct VehicleOrderInsertArgs
+    {
+        static constexpr auto command = GameCommand::vehicleOrderInsert;
+
+        VehicleOrderInsertArgs() = default;
+
+        explicit VehicleOrderInsertArgs(const registers& regs)
+            : head(EntityId(regs.di))
+            , rawOrder(static_cast<uint64_t>(regs.cx) << 32 | regs.eax)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1351]: put more brackets around this `(static_cast<uint64_t>(regs.cx) << 32) | regs.eax`

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 18]
```diff
@@ -1294,6 +1719,36 @@ namespace OpenLoco::GameCommands
...
+            , bufferIndex(regs.ax)
+        {
+            memcpy(newName, &regs.edx, 4);
+            memcpy(newName + 4, &regs.ebp, 4);
+            memcpy(newName + 8, &regs.edi, 4);
+        }
+
+        TownId townId;
+        uint16_t bufferIndex;
+        char newName[32 * 16];
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:1738]: 37

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 19]
```diff
@@ -1522,6 +1998,27 @@ namespace OpenLoco::GameCommands
...
+            : head(EntityId(regs.di))
+        {
+        }
+
+        EntityId head;
+
+        explicit operator registers() const
+        {
+            registers regs;
+            regs.bl = Flags::apply | Flags::flag_3 | Flags::flag_6;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:2016]: remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 20]
```diff
@@ -1608,6 +2105,25 @@ namespace OpenLoco::GameCommands
...
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:2121]: remove

File: src/OpenLoco/src/GameCommands/GameCommands.h [hunk 21]
```diff
@@ -1753,6 +2462,29 @@ namespace OpenLoco::GameCommands
...
+        {
+        }
+
+        EntityId head;
+        uint32_t orderOffset;
+
+        explicit operator registers() const
+        {
+            registers regs;
+            regs.cx = enumValue(head);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/GameCommands/GameCommands.h:2482]: regs.di

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 2 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:786]: only the isPrimary initialization, right?

Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:786]: and if setColourMode is not 0?

Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:786]: ah, i use the value from al, right?

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 4 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:783]: Sorry, my bad

Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:783]: Couldn´t we extend the structs from an abstract one to require the command member be required? Would help in these cases

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 5 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:991]: Sorry, how I should do then? Just remove or move to somewhere else?

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 6 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:999]: also, just remove or do something else?

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 5 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:991]: Ahhh, sorry. Just remembered that the point is just to remove these.

File: src/OpenLoco/src/GameCommands/GameCommands.h (hunk 15 above)
Comment (reinaldorauch) [reply src/OpenLoco/src/GameCommands/GameCommands.h:1249]: Sorry, copypasta

**Comment by reinaldorauch** [comment]
Hmm, rebased and built the solution and it's not loading my saves :(

**Comment by AaronVanGeffen** [comment]
> Hmm, rebased and built the solution and it's not loading my saves :(

That's odd, because the new argument structs introduced in this PR aren't replacing any of the `do_` functions yet. Do the saves load fine on master?

**Comment by reinaldorauch** [comment]
@AaronVanGeffen yeah, tried on master too and it´s the same issue

**Comment by AaronVanGeffen** [comment]
> Sure, we could do in parts but i think (if its ok) we could merge this as it will not be called directly right now and make the small prs of actually using the structs

I would prefer not to do that, to be honest. Better to branch off from master again and introduce structs one at a time instead, imo.

**Comment by reinaldorauch** [comment]
Closing this in favor of individual PRs for each game command refactor

**Pull request closed without merging**